sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/reflect/DoFnInvoker.java

```diff
@@ -95,7 +96,8 @@ void invokeOnTimer(
   void invokeSplitRestriction(ArgumentProvider<InputT, OutputT> arguments);
 
   /** Invoke the {@link TruncateRestriction} method on the bound {@link DoFn}. */
-  void invokeTruncateRestriction(ArgumentProvider<InputT, OutputT> arguments);
+  <RestrictionT> Optional<RestrictionT> invokeTruncateRestriction(
```

Comment:
The reason why I though it would make sense to return Optional was that I was hoping that it would support the case when RestrictionT being null is a valid use case.

This doesn't turn out to work with Optional since it doesn't support null values. Would you rather have this API return null like Python returning None for consistency? (null meaning that there is nothing meaningful to process)

Reply:
My intent was to use `Optional.empty()` to represent there is no more work. So we could have 2 options:
1. Use `Optional.empty()` as a notion of no more remaining work and use `Optional.of(value)` as a  notion of bounded remaining work.
2. Change the function signature to `@Nullable RestrictionT truncateRestriction()`, where `null` means no more work.

I don't have a strong preference over these 2 options. But 2 is more consistent with python.